text/0000-layer-metadata-files.md

```diff
@@ -0,0 +1,176 @@
+├── group.toml
+├── plan.toml
+├── layers
+│   └── <escaped_buildpack_id> (`CNB_BP_LAYERS_DIR` )
+│       ├── <layer>.toml
+│       └── <layer>
+└── config
+    ├── metadata.toml
+    └── <escaped_buildpack_id> (`CNB_BP_CONFIG_DIR`  )
+        ├── build.toml
+        ├── launch.toml
+        └── store.toml
+```
+
+The top level `<output-dir>` would be the mounted volume that contains the output from buildpacks. The `<layers>` sub-directory would contains the layer contributions by each buildpack under an `<escaped-buildpack-id>` directory. This directory would be passed to the buildpack via the environment variable `CNB_BP_LAYERS_DIR`. Similarly any config files would reside in the `<config>`, under a sub-directory for each buildpack. This directory would be passed to the buildpack via the environment variable `CNB_BP_CONFIG_DIR`.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle phases that reference the `<layers>` directory would now take in `<output-dir>` directory instead. The behavior with respect to the various metadata files and the lifecycle would remain the same, however the `analysis`, `restore`, `build` and `export` phases would have to be updated to look for the various layer directories inside the `layers` and `config` sub-folder in the `<output-dir>`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+This will break backwards compatibility for numerous buildpacks that don't use a set of bindings like `libcnb`. This would also result in a one-time loss of layer caches for app images. It would also result in longer path names. We would also have to re-work the export logic to account for the export of the various files which are partially present in the `config` sub-directory and the `layers` sub-directory.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+
+## Alternative 1
+
+We could add a prefix to each of the layer metadata files like `layer.`. This would allow us to re-use the cached layers but may lead to awkwardness around naming the layer metadata files - for eg. if a layer is called `layer.build` the metadata file would end up being `layer.layer.build.toml`
+
+```
+`/layers` (configurable with `CNB_LAYERS_DIR` or -layers) 
+├── group.toml
+├── plan.toml
+├── <escaped_buildpack_id> (buildpack arg `$1` )
+│   ├── build.toml
+│   ├── launch.toml
+│   ├── store.toml
+│   ├── layer.<layer>.toml
+│   └── <layer>
+└── config
+    └── metadata.toml
+```
+
+## Alternative 2
+
+We could add a prefix for each of the special metadata files and reserve it for CNB usage. For eg. `launch.toml` could be renamed to `cnb.launch.toml`. This way, we will only be prohibiting layers that have a prefix `cnb.` or a layer named `cnb`. The downsides to this alternative are the fact that the metadata files would not be usable across builds and we would still be prohibiting a class of layer names that may be valid.
```

Comment:
> the metadata files would not be usable across builds

Do you mean that if a buildpack upgrades from the old api to the new api the metadata file can't be re-used? I think the lifecycle could handle the re-naming if that's required.